### PR TITLE
fix #1980 user defined property metadata takes precedence

### LINKED ISSUE #1980: Impossible to set @ApiProperty(writable=false)
Hey there,

It appears that
`@ApiProperty(writable=false)`
is ignored when added to a property (it remains as true)

This is because the `AnnotationPropertyMetadataFactory` refuses to overwrite an existing value if the decorated `PropertyMetadataFactoryInterface` has already set it.
https://github.com/api-platform/core/blob/master/src/Metadata/Property/Factory/AnnotationPropertyMetadataFactory.php#L137-L138

It is already set here:
https://github.com/api-platform/core/blob/master/src/Bridge/Symfony/PropertyInfo/Metadata/Property/PropertyInfoPropertyMetadataFactory.php#L68

A few questions:

1. Should later running `PropertyMetadataFactory` services overwrite the one they are decorating, or do earlier ones have priority? 
2. What should be the order be of the `PropertyMetadataFactoryInterface`? Here is the current decoration order:
`CachedPropertyMetadataFactory`: `-10`
`ValidatorPropertyMetadataFactory` `20`
`AnnotationSubresourceMetadataFactory`: `30`
`SerializerPropertyMetadataFactory`: `30`
`DoctrineOrmPropertyMetadataFactory`: `40`
`AnnotationPropertyMetadataFactory`: `40`
`PropertyInfoPropertyMetadataFactory`: `40`
`InheritedPropertyMetadataFactory`: `40`
`ExtractorPropertyMetadataFactory`: `40`
3. Is it a problem that many of them have the same priority, as the order is not deterministic.

cc @soyuka @dunglas

### PR DESCRIPTION
**Pull request by soyuka**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #1980 
| License       | MIT
| Doc PR        | na

See issue.

### CODE REVIEW

**Comment by soyuka** [comment]
ping @oxan @bendavies

**Comment by oxan** [comment]
I don't think this fixes #1994, as the `SerializerPropertyMetadataFactory` still overwrites information from the annotations. I think a check needs to be added in that factory to avoid overwriting existing information. I'm not sure that won't break anything though, as it also restricts the serializer factory from overwriting data coming from Doctrine and PropertyInfo.

**Comment by oxan** [comment]
Maybe it's better to leave the serializer factory alone and lower the priority of the annotation & extractor factories so that they run last (well, before the cache). Along with this PR to let them overwrite the present metadata, that should work fine. I checked and none of the existing factories seem to read from the metadata passed to them anyway (except to avoid overwriting), so I don't think it's a problem if they don't get the metadata configured by the user.

(Actually, where I'm talking about the serializer factory, we should really think about all factories that set metadata based on heuristics. It's just that in my case the metadata I defined was overwritten by the serializer factory, but that could happen with others as well).

**Comment by soyuka** [comment]
> Maybe it's better to leave the serializer factory alone and lower the priority of the annotation & extractor factories so that they run last (well, before the cache).

Yes I think that this is safer :p. I added a commit that does this.

**Comment by oxan** [comment]
Does Symfony guarantee an order of services with the same priority? Because the `ValidatorPropertyMetadataFactory` also has priority 20, so if no order is guaranteed the validator can still overwrite user-defined metadata.

File: src/Metadata/Property/Factory/AnnotationPropertyMetadataFactory.php [hunk 1]
```diff
@@ -134,10 +134,6 @@ private function createMetadata(ApiProperty $annotation, PropertyMetadata $paren
     private function createWith(PropertyMetadata $propertyMetadata, array $property, $value): PropertyMetadata
     {
         $getter = $property[0].ucfirst($property[1]);
-        if (null !== $propertyMetadata->$getter()) {
```
Comment (dunglas) [inline src/Metadata/Property/Factory/AnnotationPropertyMetadataFactory.php]: This shouldn't be dropped. It's a BC break.

File: src/Metadata/Property/Factory/ExtractorPropertyMetadataFactory.php [hunk 2]
```diff
@@ -117,7 +117,7 @@ private function update(PropertyMetadata $propertyMetadata, array $metadata): Pr
         ];
 
         foreach ($metadataAccessors as $metadataKey => $accessorPrefix) {
-            if (null === $metadata[$metadataKey] || null !== $propertyMetadata->{$accessorPrefix.ucfirst($metadataKey)}()) {
```
Comment (dunglas) [inline src/Metadata/Property/Factory/ExtractorPropertyMetadataFactory.php]: Same here.

**Comment by bendavies** [comment]
So I'd the factory can not be changed, the only possible fix is to alter the priority?

**Comment by dunglas** [comment]
Yes. It's cleaner anyway. And it allows anybody to customize the order (by changing priorities).

**Comment by bendavies** [comment]
But the factories are still weird in that some will overwrite, but some won't. Guess this can't be changed until 3.0?

**Comment by dunglas** [comment]
None should...

**Comment by bendavies** [comment]
I'll check tomorrow but I was under the impression that wasn't the case.

**Comment by soyuka** [comment]
I'm really unsure that changing priorities will be enough to fix this...

**Comment by oxan** [comment]
Yeah, at least the `SerializerPropertyMetadataFactory` currently does overwrite existing data. Since the annotation factory doesn't, only changing the priority doesn't fix this.

**Comment by teohhanhui** [comment]
@oxan

> Does Symfony guarantee an order of services with the same priority?

It should be in the order of service registration: https://github.com/symfony/symfony/blob/v4.1.0/src/Symfony/Component/DependencyInjection/Compiler/DecoratorServicePass.php#L28

**Comment by soyuka** [comment]
So I should change `SerializerPropertyMetadataFactory` so that they don't overwrite? But will it not be a BC break?

**Comment by bendavies** [comment]
that seems like it would be a BC break yes. seems like a catch 22 situation here to me.

**Comment by bendavies** [comment]
copying @Nek- nice image which shows the current state of things:

**Comment by Nek-** [comment]
Btw I can update it on demand.

**Comment by soyuka** [comment]
Anyway, if you guys need a resolution to this issue, you can always add your own `PropertyMetadataFactory` that overrides everything (with the lowest priority before cache)! 

To me this whole issue looks like a wontfix because I don't see any viable solution that doesn't break the current state of things...

**Comment by teohhanhui** [comment]
@bendavies There is no "ambiguous order" :)

**Comment by bendavies** [comment]
@teohhanhui it's not my image, but anyway, it's more a "non obvious" order.

**Comment by oxan** [comment]
Is there any actual situation where autogenerated metadata overwiriting user-defined metadata is actually desired? Yes, preventing that is a BC break, but every bugfix is a BC break in a way. I can't think of any real-world situations where the current situation would be relied upon. The behaviour would only change in places where people purposefully have manually declared wrong metadata (in fact, `@ApiProperty` is barely documented, so I doubt it has much usage at all).

**Comment by oxan** [comment]
@soyuka Yes, I'm currently using a custom metadata factory to be able to override the `readableLink` and `writableLink` properties, but it's ugly and it'd be nice to have an easier solution (for future reference, code is here: https://gist.github.com/oxan/5a9ba44e72bbeae5fa1aee7d0be0b337)

**Comment by bendavies** [comment]
agree with @oxan really.

**Comment by Nek-** [comment]
I think a failing test case is needed to show how obviously wrong is "state of things". I'll do it next week if nobody is faster than me.

**Comment by soyuka** [comment]
In the mean time we could always introduce a new annotation that overrides metadata (like the one from @oxan), feels dirty but at least it's non-breaking.

File: src/Metadata/Property/Factory/SerializerPropertyMetadataFactory.php [hunk 3]
```diff
@@ -117,7 +114,7 @@ private function transformLinkStatus(PropertyMetadata $propertyMetadata, array $
         $relatedClass = $type->isCollection() && ($collectionValueType = $type->getCollectionValueType()) ? $collectionValueType->getClassName() : $type->getClassName();
 
         if (null === $relatedClass) {
-            return $propertyMetadata;
+            return $propertyMetadata->withReadableLink(true)->withWritableLink(true);
```
[the remark this answers is not in the document]
Comment (teohhanhui) [reply src/Metadata/Property/Factory/SerializerPropertyMetadataFactory.php]: Yeah, I don't think this is correct.

[the remark this answers is not in the document]
Comment (soyuka) [reply src/Metadata/Property/Factory/SerializerPropertyMetadataFactory.php]: I'm totally :+1: to remove this but this has some impacts down the road.
Should we default these to `true` or should we test for `null || true` in the normalizers?

[the remark this answers is not in the document]
Comment (teohhanhui) [reply src/Metadata/Property/Factory/SerializerPropertyMetadataFactory.php]: Hmm... Now I see we were already doing this from before. :scream: I guess let's keep it then. Haha...

**Comment by soyuka** [comment]
@dunglas wdyt?

**Comment by bendavies** [comment]
🚢

**Pull request merged**